lib/Sema/CSDiag.cpp

```diff
@@ -6508,7 +6510,104 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
         }
         return true;
       }
-      
+
+      if (inferredArgCount == 1 && actualArgCount > 1) {
+        // Let's see if inferred argument is actually a tuple inside of Paren.
+        if (auto *argTupleTy = inferredArgType->getAs<TupleType>()) {
+          // Looks like the number of closure parameters matches number
+          // of inferred arguments, which means we can we can emit an
+          // error about an attempt to make use of tuple splat or tuple
+          // destructuring and provide a proper fix-it.
+          if (argTupleTy->getNumElements() == actualArgCount) {
+            // In case of implicit parameters e.g. $0, $1 we
+            // can't really provide good fix-it because
+            // structure of parameter type itself is unclear.
+            for (auto *param : params->getArray()) {
+              if (param->isImplicit()) {
+                diagnose(params->getStartLoc(),
+                         diag::closure_tuple_parameter_destructuring_implicit,
+                         argTupleTy);
+                return true;
+              }
+            }
+
+            auto diag = diagnose(params->getStartLoc(),
+                                 diag::closure_tuple_parameter_destructuring,
+                                 argTupleTy);
+            Type actualArgType;
+            if (auto *actualFnType = CE->getType()->getAs<AnyFunctionType>())
+              actualArgType = actualFnType->getInput();
+
+            auto &sourceMgr = CS->getASTContext().SourceMgr;
+            auto paramRange = CharSourceRange(
+                sourceMgr, params->getStartLoc(),
+                Lexer::getLocForEndOfToken(sourceMgr, params->getEndLoc()));
+
+            SmallString<64> fixIt;
+            llvm::raw_svector_ostream OS(fixIt);
+
+            // If this is multi-line closure we'd have to insert new lines
+            // in the suggested 'let' to keep the structure of the code in
+            // tact, otherwise just use ';' to keep everything on the same
+            // line.
+            bool isMultiLineClosure = false;
+            if (auto *body = CE->getBody()) {
+              auto inLine = sourceMgr.getLineNumber(CE->getInLoc());
+              auto bodyLine = sourceMgr.getLineNumber(body->getStartLoc());
+              isMultiLineClosure = bodyLine > inLine;
+            }
+
+            std::string parameter;
+            auto paramText = sourceMgr.extractText(paramRange);
+            // If parameter is not enclosed in parenthesis,
+            // let's add them since we are forming a tuple.
+            if (!paramText.startswith("(")) {
+              parameter = "(" + paramText.str() + ")";
+            } else {
+              parameter = paramText;
+            }
+
+            // If at least one of the parameters has explicitly
+            // declared type let's turn fix-it into N 'let' expressions
+            // to be able to specify it.
```

Comment:
That's correct, the proper way to write that is:
``
let (x, y): (Int, Double) = (0,0)
``

Writing:
``
let (x: Int, y: Double) = (0, 0)
``
creates a pattern_tuple with element names x and y.